File: ts/src/parse.ts
import { parseHTML } from "linkedom";

import * as fs from "fs";

type Message = {
  id: string;
  fromEmail: string;
  fromName: string | null;
  htmlBody: string;
};

function extractLinks(html: string): string[] {
  const { document } = parseHTML(html);
  const links: string[] = [];

  document.querySelectorAll("a").forEach((link) => {
    const target = link.getAttribute("href");
    if (target && target.startsWith("http")) {
      links.push(target);
    }
  });

  return links;
}

(async () => {
  const startTime = new Date().getTime();

  const messages: Message[] = JSON.parse(
    fs.readFileSync("../messages.json", "utf-8")
  );
  const allLinks: string[] = [];

  messages.forEach((message) => {
    if (message.htmlBody) {
      const links = extractLinks(message.htmlBody);
      if (links.length > 0) {
        allLinks.push(...links);
      }
    }
  });

  fs.writeFileSync("./links.txt", allLinks.join("\n"));

  const endTime = new Date().getTime();
  console.log(`time: ${endTime - startTime}ms`);
})();

File: go/parse.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"strings"
	"time"

	"github.com/PuerkitoBio/goquery"
)

type Message struct {
	Id        string `json:"id"`
	FromEmail string `json:"fromEmail"`
	FromName  string `json:"fromName"`
	HtmlBody  string `json:"HtmlBody"`
}

func ExtractLinks(html string) []string {
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(html))
	if err != nil {
		log.Fatal(err)
	}

	var links []string

	doc.Find("a").Each(func(i int, s *goquery.Selection) {
		target, exists := s.Attr("href")
		if exists {
			links = append(links, target)
		}
	})

	return links
}

func main() {
	start := time.Now()

	content, err := ioutil.ReadFile("../messages.json")
	if err != nil {
		fmt.Println(err)
	}

	var payload []Message
	err = json.Unmarshal(content, &payload)
	if err != nil {
		log.Fatal("Error during Unmarshal(): ", err)
	}

	var alllinks []string

	for _, value := range payload {
		alllinks = append(alllinks, ExtractLinks(value.HtmlBody)...)
	}

	ioutil.WriteFile("links.txt", []byte(strings.Join(alllinks, "\n")), 0644)

	duration := time.Since(start)
	fmt.Printf("time: %dms\n", duration.Milliseconds())
}
